### ISSUE TITLE
[HOLD for payment 2023-10-24] [$500] Web - App keeps both right click LHN popup and delete message popup open together

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#28324]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the app
2. Open any report
3. Focus on composer
4. Right on any report in LHN to open right click LHN popup
5. Type anything and press enter to send
6. Click up arrow to edit the sent message
7. Remove all content using backspace and click enter
8. Observe that app does not close right click LHN popup and opens delete popup too along with it.

## Expected Result:
App should close existing open popup on opening any other popup in the app

## Actual Result:
App keeps both right click LHN popup and delete popup open together

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.74-2

**Reproducible in staging?:** Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/5ccbf5dd-6ea5-4189-8f02-b70cf1c7f501

https://github.com/Expensify/App/assets/115492554/e52a1a16-5779-410f-b628-b2154d9d7a8c

https://github.com/Expensify/App/assets/115492554/8521de3b-5399-440e-96d4-4d9dcce00b95

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Dhanashree-Sawant

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1695654750958139

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b7d70b5aeb6fba4a</li>
        <li>Upwork Job ID: 1707062073662357504</li>
        <li>Last Price Increase: 2023-09-27</li>
<li>Automatic offers: </li>
<ul>
<li>jjcoffee | Reviewer | 27017549</li>
<li>Dhanashree-Sawant | Reporter | 27017552</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @dylanexpensify</details>

**Comment by ikevin127** [comment]
# Proposal
## Please re-state the problem that we are trying to solve in this issue
App keeps both right click LHN popup and delete popup open together.

## What is the root cause of that problem?
The problem is that within the `PopoverReportActionContextMenu.js` component, the LHN popup is not being closed before the delete message popup is opened.

## What changes do you think we should make in order to solve the problem?
Within the `PopoverReportActionContextMenu.js` component I called the `hideContextMenu` function within the `showDeleteModal` function which makes sure the LHN popup is closed before the delete message popup is being opened:

```diff
   const showDeleteModal = (reportID, reportAction, shouldSetModalVisibility = true, onConfirm = () => {}, onCancel = () => {}) => {
        onCancelDeleteModal.current = onCancel;
        onComfirmDeleteModal.current = onConfirm;

        reportIDRef.current = reportID;
        reportActionRef.current = reportAction;

-        setShouldSetModalVisibilityForDeleteConfirmation(shouldSetModalVisibility);
-        setIsDeleteCommentConfirmModalVisible(true);
+        hideContextMenu(() => {
+            setShouldSetModalVisibilityForDeleteConfirmation(shouldSetModalVisibility);
+            setIsDeleteCommentConfirmModalVisible(true);
+        });
    };
```

## What alternative solutions did you explore? (Optional)
N/A

**Comment by astrohunter62** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
The LHN does not receive focus when clicking on it.
### What is the root cause of that problem?
In the OptionRowLHN, default actions are prevented when the mouse is clicked.
``` js
onMouseDown={(e) => {
  if (!e) {
      return;
  }
  // Prevent losing Composer focus
  e.preventDefault();
}}
```
### What changes do you think we should make in order to solve the problem?
We can re-enable the default actions by removing the code block above.

### Results:

[28324(safari).webm](https://github.com/Expensify/App/assets/136574831/3c0c09e3-b7de-4561-be6a-eb26aff46c20)

[28324(chrome).webm](https://github.com/Expensify/App/assets/136574831/7ea008a7-4511-44ff-80ae-b0de0711b917)

###  What alternative solutions did you explore? (Optional)
If you prefer not to have this behavior, you have the option to call the hideContextMenu function within the showDeleteModal function to ensure the LHN popup is closed before opening the delete message popup.

**Comment by kaushiktd** [comment]
### Please re-state the problem that we are trying to solve in this issue
 The LHN popup is still open after opening the delete message popup.

### What is the root cause of that problem?
The root cause of the issue is in the file PopoverReportActionContextMenu.js, within the showDeleteModel function. The issue occurs because the delete modal is opened without closing it first, resulting in both the delete modal and any previously open popups remaining visible simultaneously.

### What changes do you think we should make in order to solve the problem?

Add this line of code which closes LHN popup before opening the delete popup
https://github.com/Expensify/App/blob/main/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L264

```
if (this.state.isPopoverVisible) {
            this.hideContextMenu();
     }
```

### Video:- 
https://drive.google.com/file/d/1ifmq4_JOlrPy7gMaT7ahxPTgGQuQRFMl/view?usp=sharing

**Comment by dylanexpensify** [comment]
@jjcoffee mind giving these proposals a review today?

**Comment by jjcoffee** [comment]
Oh I'm sorry, not sure how this one slipped past me! Reviewing today!

**Comment by jjcoffee** [comment]
Looks like the correct RCA here is in @astrohunter62's [proposal](https://github.com/Expensify/App/issues/28324#issuecomment-1738077181), as we are always keeping focus on the composer when we should be defocusing since we're opening the right click menu. 

It looks like this is a regression from [this PR](https://github.com/Expensify/App/pull/21552), which was attempting to ensure that the composer refocuses when clicking an already-open chat in the LHN. So, we want to keep that behaviour but _right_ clicking LHN should lose focus from composer.

Could you update your proposal based on the above @astrohunter62? Also a side-note, can you please [link](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet) to the code snippet you mention in your proposal - makes it much easier to review!

The other two proposals are decent, but miss out because of not identifying the correct RCA.

:ribbon::eyes::ribbon: C+ reviewed

**Comment by astrohunter62** [comment]
@jjcoffee
I updated my proposal https://github.com/Expensify/App/issues/28324#issuecomment-1738077181

**Comment by jjcoffee** [comment]
@astrohunter62 Perfect, thanks! I think code block you're proposing to remove would cause a regression from [this PR](https://github.com/Expensify/App/pull/21552) though.

**Comment by astrohunter62** [comment]
@jjcoffee Thank you.
If you prefer not to remove it, you can alternatively insert ``DomUtils.blurActiveElement();`` into the ``onSecondaryInteraction`` within ``OptionRowLHN``. However, I believe this action is more unnecessary 
```js
onSecondaryInteraction={(e) => {
    showPopover(e);
    DomUtils.blurActiveElement();  // add
}}
```
But I think the previous solution(https://github.com/Expensify/App/issues/28324#issuecomment-1738077181) is better than this one.

**Comment by jjcoffee** [comment]
@astrohunter62 Sorry I don't see how your [proposal](https://github.com/Expensify/App/issues/28324#issuecomment-1738077181) addresses the issue in [that PR](https://github.com/Expensify/App/pull/21552). I was thinking more to limit the `onMouseDown` to return early for right-clicks.

**Comment by astrohunter62** [comment]
@jjcoffee Did you see my alternative solution(https://github.com/Expensify/App/issues/28324#issuecomment-1746702357)?

**Comment by jjcoffee** [comment]
@astrohunter62 Yes, I still don't see how it's related. Can you explain more? Did you look at [the PR](https://github.com/Expensify/App/pull/21552) I mentioned?

**Comment by astrohunter62** [comment]
@jjcoffee 
By using `e.preventDefault()`; within the `onMouseDown`, prevent the input from losing focus when clicking on the same chat in LHN.
 In my alternative solution, I don't remove it, and the composer loses focus when the Popover is displayed. This solution doesn't cause a regression from https://github.com/Expensify/App/pull/21552

**Comment by jjcoffee** [comment]
@astrohunter62 Ah I get you now, thanks! I think I still prefer an early return in `onMouseDown` as `onSecondaryInteraction` runs for long presses as well as right-clicks. We can deal with this on the PR though, so you can go ahead and raise it!

**Comment by astrohunter62** [comment]
The [PR](https://github.com/Expensify/App/pull/28846) is ready to review. cc @jjcoffee @dylanexpensify

**Comment by jjcoffee** [comment]
@dylanexpensify PR is under review.

**Comment by jjcoffee** [comment]
@dylanexpensify Looks like this actually got fixed by another PR https://github.com/Expensify/App/pull/28953. Is there any compensation due in that case?

**Comment by jjcoffee** [comment]
@dylanexpensify Actually that PR fixes it in a way that causes [issues on tablets](https://github.com/Expensify/App/pull/28846#issuecomment-1755797888). Is that worth us continuing with our PR to fix do you think?

**Comment by jjcoffee** [comment]
So just to update, that [other PR](https://github.com/Expensify/App/pull/28953)'s fix isn't really complete as beyond it not working on tablets it also doesn't refocus on the composer if clicking onto the same chat after closing the context menu (see [here](https://github.com/Expensify/App/pull/28846#issuecomment-1757394162)), which is a regression from https://github.com/Expensify/App/pull/21552. So we're proceeding with our fix here.

**Comment by astrohunter62** [comment]
@jjcoffee
I sent a proposal on Upwork, but I haven't received any offer.
When can I expect an offer?

**Comment by jjcoffee** [comment]
@astrohunter62 For your first job you will usually get hired on Upwork once payment is ready to be made (which happens once the PR hits production + a 7 day regression period). Then for future issues you should start automatically receiving Upwork offers as soon as you get assigned to the GH issue.

**Comment by jjcoffee** [comment]
@astrohunter62 Are you able to handle the [above](https://github.com/Expensify/App/issues/29746) regression?

**Comment by astrohunter62** [comment]
Yes.

**Comment by astrohunter62** [comment]
@jjcoffee 
Can you review this code?
```js

onSecondaryInteraction={(e) => {
    showPopover(e);
    // Ensure that we blur the composer when opening context menu, so that only one component is focused at a time
    if (DomUtils.getActiveElement()) {
        DomUtils.getActiveElement().blur();
    }
}}
```

**Comment by jjcoffee** [comment]
@astrohunter62 Please raise a PR, I will review there.

**Comment by astrohunter62** [comment]
Can I make PR?

**Comment by jjcoffee** [comment]
@astrohunter62 Yes, make sure to link the [regression issue](https://github.com/Expensify/App/issues/29746) as well as this one.

**Comment by astrohunter62** [comment]
@jjcoffee  Can you review this [PR](https://github.com/Expensify/App/pull/29752)?

**Comment by jjcoffee** [comment]
Fix is [merged](https://github.com/Expensify/App/pull/29752#event-10678081317)!

**Comment by astrohunter62** [comment]
@jjcoffee Hi,
When can I get an offer?

**Comment by jjcoffee** [comment]
- The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/21552 (also https://github.com/Expensify/App/pull/28953 which attempted to fix it)
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/21552#discussion_r1365671613 (original PR) and https://github.com/Expensify/App/pull/28953#issuecomment-1780666536 (attempted fix PR)
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **N/A - just needed more thorough testing**
- Determine if we should create a regression test for this bug. **Yes**
- If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

**Regression test proposal**

1. Open any report and click on the composer to focus it.
2. Right click on any report in LHN to open the context menu.
3. Verify that the composer loses focus.
4. Click on the same chat in LHN and verify that the composer regains focus.

Do we agree 👍 or 👎

**Comment by dylanexpensify** [comment]
Payment summary:

- Reporter: @dhanashree-sawant $50
- Contributor: @astrohunter62 $250
- Contributor+: @jjcoffee $250

Please request money or apply to upwork job!

**Comment by dylanexpensify** [comment]
@astrohunter62 sent offer!

@dhanashree-sawant and @jjcoffee payment sent!

**Comment by astrohunter62** [comment]
@dylanexpensify 
I have sent my work. Could you approve it?

**Comment by dylanexpensify** [comment]
@jjcoffee ah I requested half back for you to refund! It was easier than sending a new offer, waiting for you to accept, etc. haha! @astrohunter62 looking now!

**Comment by dylanexpensify** [comment]
payments sent!

**Comment by jjcoffee** [comment]
@dylanexpensify Oooh, interesting, for some reason Upwork didn't send a notification with the refund request. Refunded now!
